File: python3/day_28_Reconstruct Itinerary.py
class Solution:

    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        def dfs(graph, u, res):
            while(len(graph[u])):    
                dfs(graph, graph[u].pop(), res)

            res.append(u)

        graph = defaultdict(list)
        res = []

        for ticket in tickets:
            graph[ticket[0]].append(ticket[1])
        
        for k in graph:
            graph[k].sort(reverse=True)

        u = "JFK"
        dfs(graph, u, res)
        res.reverse()
        return res

File: cpp/june/day_28_Reconstruct Itinerary.cpp
class Solution {
public:
    void dfs(unordered_map<string, vector<string>> &graph, string u, vector<string> &res){
        
        for(auto it = graph[u].begin(); it != graph[u].end(); ){
            string v = *it;
            graph[u].erase(it);
            dfs(graph, v, res);
        }
        res.push_back(u);
    }
    
    vector<string> findItinerary(vector<vector<string>>& tickets) {
        unordered_map<string, vector<string>> graph;
        vector<string> res;
        int i, j;
        string u, v;
        for(i = 0; i < tickets.size(); i++){
            graph[tickets[i][0]].push_back(tickets[i][1]);
        }
        
        for(auto it = graph.begin(); it != graph.end(); it++){
            sort(it->second.begin(), it->second.end());
        }

        u = "JFK";
        dfs(graph, u, res);
        reverse(res.begin(), res.end());
        return res;
    }
};
